### ISSUE TITLE
UI Test Process crashing, seems to be related to leakCanary

### ISSUE THREAD

**Issue by npark-mozilla** [mozilla-mobile/focus-android#2068]
Locally the tests (For example, PageVisitTest) started failing with 'Process Crashed' error, and submitting the test on Google Testcloud shows following stacktrace.

java.lang.NoSuchMethodError: No direct method <init>(Lcom/squareup/leakcanary/WatchExecutor;Lcom/squareup/leakcanary/DebuggerControl;Lcom/squareup/leakcanary/GcTrigger;Lcom/squareup/leakcanary/HeapDumper;Lcom/squareup/leakcanary/HeapDump$Listener;Lcom/squareup/leakcanary/ExcludedRefs;)V in class Lcom/squareup/leakcanary/RefWatcher; or its super classes (declaration of 'com.squareup.leakcanary.RefWatcher' appears in /data/app/org.mozilla.focus.debug-1/base.apk:classes32.dex)

FATAL EXCEPTION: main
Process: org.mozilla.focus.debug, PID: 6468
java.lang.NoSuchMethodError: No direct method <init>(Lcom/squareup/leakcanary/WatchExecutor;Lcom/squareup/leakcanary/DebuggerControl;Lcom/squareup/leakcanary/GcTrigger;Lcom/squareup/leakcanary/HeapDumper;Lcom/squareup/leakcanary/HeapDump$Listener;Lcom/squareup/leakcanary/ExcludedRefs;)V in class Lcom/squareup/leakcanary/RefWatcher; or its super classes (declaration of 'com.squareup.leakcanary.RefWatcher' appears in /data/app/org.mozilla.focus.debug-1/base.apk:classes32.dex)
	at com.squareup.leakcanary.RefWatcherBuilder.build(RefWatcherBuilder.java:88)
	at com.squareup.leakcanary.AndroidRefWatcherBuilder.buildAndInstall(AndroidRefWatcherBuilder.java:57)
	at com.squareup.leakcanary.LeakCanary.install(LeakCanary.java:42)
	at org.mozilla.focus.FocusApplication.onCreate(FocusApplication.java:36)
	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1024)
	at android.support.test.runner.MonitoringInstrumentation.callApplicationOnCreate(MonitoringInstrumentation.java:382)
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5403)
	at android.app.ActivityThread.-wrap2(ActivityThread.java)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1545)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:154)
	at android.app.ActivityThread.main(ActivityThread.java:6119)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)

**Comment by npark-mozilla** [comment]
Seems like this is the commit when it was introduced:
https://github.com/mozilla-mobile/focus-android/commit/5bb96d68a874b6f03c87a4fcd28e11cb041b290e

**Comment by npark-mozilla** [comment]
But it looks like BB is running tests fine - so maybe there is something that I missed? I just built it on usual focusWebviewDebug mode and ran tests..
https://dashboard.buddybuild.com/apps/584f67f6f3d6eb01000842d6/build/5a67a09f3fb6da000182baa8#uitests

**Comment by npark-mozilla** [comment]
This seemed to affect both Android Studio IDE, and ./gradlew command. I can't get it to work even after recloning the directory, and delete Andriod Studio cache. @siddware thought?

**Comment by npark-mozilla** [comment]
Just discovered that after changing https://github.com/mozilla-mobile/focus-android/commit/5bb96d68a874b6f03c87a4fcd28e11cb041b290e#diff-39e7d8c00954e920b98e7636f0ac30b2R184
to "releaseCompile "com.squareup.leakcanary:leakcanary-android-no-op:$leakcanary_version"
as shown in leakCanary github page, the tests are back running.  @siddware was there a specific reason that 'compile' cmd is used rather than 'releaseCompile'?

**Comment by npark-mozilla** [comment]
Ah I see now, if releaseCompile is used, it'll break unit tests.

**Comment by siddware** [comment]
@npark-mozilla Sorry about the delay with getting back to you. To add some further information, the ticket specified that leak canary should only be enabled for debug builds, so compile was used to include all the other types instead of just release. I will look into this a bit more and will get back to you

**Comment by npark-mozilla** [comment]
@siddware thanks! if I use '---Compile' command for for each of other types as in #2071, I see unit test failures... so kinda curious whether our unit tests need to be modified.  But if it turns out that I'm missing something in my local setup, please let me know.  I think the reason bb is passing is that it sends out two separate gradle build commands, but I could be wrong.

**Comment by siddware** [comment]
@npark-mozilla To add to the mystery, I cannot reproduce this issue at all. I ran PageVisitTest and it passed, and then I ran all the tests in the activity folder, only one failed and even that passed when ran in isolation. Out of two tests in privacy folder, one failed and kept failing even when ran in isolation. All the tests in screenshots folder passed. 
As far as Unit tests go, 5 failed (from 2 files) out of 331. And these failed when ran in isolation as well. But I don't think any of these errors are due to LeakCanary. I've attached the logs, please let me know if you need more information or want me to try something else. 

Test conditions:
I pulled in latest changes and branched off master. All instrumentation tests were run on a Oneplus one running Lineage OS (Android 7.1.1). The build variant is FocusWebviewDebug. Nothing estoeric about Android Studio, latest version, running on mac OS 10.13.2.
[test-logs-focus.zip](https://github.com/mozilla-mobile/focus-android/files/1693294/test-logs-focus.zip)

**Comment by npark-mozilla** [comment]
@siddware it seems that I might have a path issue.  I wiped the folder, cloned at /Users/mozilla/focus-android/ and /Users/mozilla/GitRepo/focus-android separately, and ran. the former ran fine, the latter failed, saying that it can't find the leakcanary method.  I actually am not sure how this is possible, since nothing seems to be hardcoded, but I will check.  Thank you for checking on this!

**Comment by npark-mozilla** [comment]
here’s one additional thing I found:  my focus-android folder was under ~/GitRepo originally, which exhibited this crash.  (even just running the app will cause the crash).  
I renamed GitRepo folder to 'GitRepoo' and rebuilt/reran the app. the crash still happens.
Then, I renamed GitRepo folder to 'abc' and rebuiit/reran the app. the crash is no longer reproducible.
Similarly,
GitRep -> crash happens
gitrepo -> crash doesn't happen
GITREPO -> crash doesn't happen
AbCd -> crash happens

So I'm thinking, maybe the mix of cases is somehow confusing the gradle build?
